### PR TITLE
[iam policy] EOS-25962: Attach policy to user

### PR DESCRIPTION
**Pull request by shrimetta**
Signed-off-by: Shri Metta <shri.metta@seagate.com>

# Problem Statement
- Attach policy to user

# Design
-  https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/706805920/UML+Diagrams

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [x] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [x] Interface change (if any) are documented
- [x] Side effects on other features (deployment/upgrade)
- [x] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [x] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Review by shalakadharap — APPROVED** [review]
Looks fine. Fix codacy issues for this PR
and post the link of jenkins job as well.

**Review by bkirunge7 — APPROVED** [review]

File: auth/server/src/main/java/com/seagates3/dao/ldap/UserPolicyLdapStore.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        String.format("%s=%s,%s=%s,%s=%s,%s=%s,%s", LDAPUtils.USER_ID,
+                      user.getId(), LDAPUtils.ORGANIZATIONAL_UNIT_NAME,
+                      LDAPUtils.USER_OU, LDAPUtils.ORGANIZATIONAL_NAME,
+                      user.getAccountName(), LDAPUtils.ORGANIZATIONAL_UNIT_NAME,
+                      LDAPUtils.ACCOUNT_OU, LDAPUtils.BASE_DN);
+
+    if (policy.getPolicyId() != null) {
+      List<String> userPolicyIds = user.getPolicyIds();
+      userPolicyIds.add(policy.getPolicyId());
+      String[] policyIds = new String[userPolicyIds.size()];
```
Comment (shalakadharap) [inline auth/server/src/main/java/com/seagates3/dao/ldap/UserPolicyLdapStore.java:43]: Need to increment attachment_count for policy when you are attaching a policy to a user.
Plz handle it.
Exactly opposite for detach- decrement the counter

Comment (shrimetta) [reply auth/server/src/main/java/com/seagates3/dao/ldap/UserPolicyLdapStore.java]: Done

File: auth/server/src/main/java/com/seagates3/dao/ldap/UserPolicyLdapStore.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    return userModList;
+  }
+
+ private
+  ArrayList getPolicyModListForAttachDetachPolicy(Policy policy,
+                                                  boolean isAttach) {
+    String newAttachmentCount =
+        isAttach ? String.valueOf(policy.getAttachmentCount() + 1)
+                 : String.valueOf(policy.getAttachmentCount() - 1);
+    ArrayList policyModList = new ArrayList();
```
Comment (shalakadharap) [inline auth/server/src/main/java/com/seagates3/dao/ldap/UserPolicyLdapStore.java:125]: why this list is required?

File: auth/server/src/main/java/com/seagates3/dao/ldap/UserPolicyLdapStore.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+                         policyName, LDAPUtils.ORGANIZATIONAL_UNIT_NAME,
+                         LDAPUtils.POLICY_OU, LDAPUtils.ORGANIZATIONAL_NAME,
+                         accountName, LDAPUtils.ORGANIZATIONAL_UNIT_NAME,
+                         LDAPUtils.ACCOUNT_OU, LDAPUtils.BASE_DN);
+  }
+
+ private
+  ArrayList getUserModListForAttachDetachPolicy(User user, Policy policy,
+                                                boolean isAttach) {
+    ArrayList userModList = new ArrayList();
```
Comment (shalakadharap) [inline auth/server/src/main/java/com/seagates3/dao/ldap/UserPolicyLdapStore.java:104]: why userModList required?

File: auth/server/src/main/java/com/seagates3/dao/ldap/UserPolicyLdapStore.java (hunk 2 above)
Comment (shrimetta) [reply auth/server/src/main/java/com/seagates3/dao/ldap/UserPolicyLdapStore.java]: Removed

File: auth/server/src/main/java/com/seagates3/dao/ldap/UserPolicyLdapStore.java (hunk 3 above)
Comment (shrimetta) [reply auth/server/src/main/java/com/seagates3/dao/ldap/UserPolicyLdapStore.java]: Removed

**Comment by shrimetta** [comment]
Pre Merge test successful.
http://eos-jenkins.colo.seagate.com/job/S3server/job/S3Server-PreMerge-Test/675/

**Pull request merged**